README
Dist::Zilla::Plugin::Prereqs::SyncVersions

This module exists to pose mostly as a workaround for potential bugs in
downstream tool-chains.

Namely, "CPAN.pm" is confused when it sees:

    runtime.requires : Foo >= 5.0
    test.requires    : Foo >= 6.0

It doesn't know what to do.

This is an easy enough problem to solve if you're using "[Prereqs]"
directly, and "[AutoPrereqs]" already does the right thing, but it gets
messier when you're working with plugins that inject their own
prerequisites
<https://github.com/dagolden/Path-Tiny/commit/c620171db96597456a182ea608
8a24d8de5debf6>

So this plugin will homogenize dependencies to be the same version in
all phases which infer the dependency, matching the largest one found,
so the above becomes:

    runtime.requires : Foo >= 6.0
    test.requires    : Foo >= 6.0

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Dist-Zilla-Plugin-Prereqs-SyncVersions

0.003003 2017-03-07T15:55:00Z 4ba8f17

0.003002 2017-03-07T15:53:19Z 8f10a03
 - Bugfix: Avoid test failures under -Ddefault_inc_excludes_dot
 - Remove use of ConfigDumper util
 - Remove use of Test::KENTNL/dztest in tests

 [Dependencies::Stats]
 - Dependencies changed since 0.003001, see misc/*.deps* for details
 - configure: (suggests: +1)
 - develop: +5 ↑4 -1 (suggests: ↑2)
 - runtime: -1
 - test: ↓1 -1

0.003001 2015-02-04T09:48:39Z ae594be
 [Dependencies::Stats]
 - Dependencies changed since 0.003000, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +10 ↑2 -2 (recommends: +1, suggests: ↑2)
 - runtime: ↓1 -1 (recommends: +1)
 - test: +1 (recommends: ↑1 ↓1)

 [Meta]
 - Add x_deprecated field

0.003000 2014-09-11T04:07:09Z
 [00 Minor]
 - use dump_config
 - metadata -> author = kentnl
 - Add tests

 [Dependencies::Stats]
 - Dependencies changed since 0.002001, see misc/*.deps* for details
 - develop: +2 (suggests: ↑1)
 - runtime: +1
 - test: +2 (recommends: ↑1)

 [Tests]
 - Basic functionality tests added.
 - Tests confirmed to keep behaviour consistent on 4.x

0.002001 2014-08-16T20:14:31Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - tests updated.
 - metadata updated.
 - whitespace reduced.

 [Dependencies::Stats]
 - Dependencies changed since 0.002000, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 ↑1 (suggests: ↑2)
 - test: +1 -2 (recommends: ↑1 -1)

 [Metadata]
 - move to kentnl org.
 - show 1x contributor.

 [Misc]
 - Improve source side POD to reduce build side whitespace.

 [Tests]
 - update Test::ReportPrereqs

0.002000 2014-06-03T06:53:36Z
 [00 Major]
 - Toolkit moved to EUMM and using Soft dependencies
 - dist.ini now baked.
 - Distribution now deprecated.

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +41 ↑1 -2 (suggests: +2 -1)
 - runtime: +1 ↑1
 - test: +3 ↓1 -3 (recommends: +4)

0.001001 2013-10-12T07:07:17Z
 [00 Minor]
 - Typo in documentation ( thanks haarg ).

 [Dependencies::Changed / develop requires]
 - Dist::Zilla::PluginBundle::Author::KENTNL v1.9.0 → 2.000000

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - develop: ↑1

0.001000 2013-10-12T04:50:38Z
 - First version.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "(DEPRECATED) Homogenize prerequisites so dependency versions are consistent",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dist-Zilla-Plugin-Prereqs-SyncVersions",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Dist::Zilla::Plugin::Prereqs::SyncVersions",
  "PREREQ_PM" => {
    "Dist::Zilla::Role::PrereqSource" => 0,
    "Moose" => 0,
    "MooseX::Types::Moose" => 0,
    "strict" => 0,
    "version" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::DZil" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.003003",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Dist::Zilla::Role::PrereqSource" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Moose" => 0,
  "MooseX::Types::Moose" => 0,
  "Test::DZil" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# FILENAME: bundle_to_ini.pl
# CREATED: 02/06/14 01:48:56 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Write an INI file from a bundle

use 5.008;    #utf8
use strict;
use warnings;
use utf8;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    multi
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  open my $rcfile, '>', './perlcritic.rc' or croak 'Cant open perlcritic.rc';
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  open my $depsfile, '>', './perlcritic.deps' or croak 'Cant open perlcritic.deps';
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4007"
  },
  "configure_requires" => {
    "Module::Build" => "0.4007"
  },
  "dist_abstract" => "Homogenize prerequisites so dependency versions are consistent",
  "dist_author" => [
    "Kent Fredric <kentfredric\@gmail.com>"
  ],
  "dist_name" => "Dist-Zilla-Plugin-Prereqs-SyncVersions",
  "dist_version" => "0.001001",
  "license" => "perl",
  "module_name" => "Dist::Zilla::Plugin::Prereqs::SyncVersions",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Dist::Zilla::Role::PrereqSource" => 0,
    "Moose" => 0,
    "MooseX::Types::Moose" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "version" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.98"
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
